docs/migrations.md

```diff
@@ -0,0 +1,144 @@
+in `setup.py`. If the extension is maintained outside of the HaaS core source
+tree, make sure you pass `zip_safe=False` to `setup`.
+
+Next, to generate a template migration, execute:
+
+    haas-admin db migrate --head haas@head \
+        --branch-label ${name_of_extension_module} \
+        -m '<Summary of change>'
+
+Alembic will look at the differences between the schema in the database and
+the one derived from the HaaS source, and generate its best attempt at a
+migration script. The script will be stored in `haas/migrations/versions`.
+Copy it to the directory you chose, and change the value of `down_revision` to
+`None`.
+
+Sanity check the output; Alembic often does a good job generating scripts, but
+it should not be trusted blindly.
+
+Finally, run `haas-admin db upgrade heads` to upgrade the database according
+to your new migration.
+
+### If the extension already has migrations, or the script is for HaaS core
+
+To generate a new migration script, execute:
+
+    haas-admin db migrate --head ${branch_name}@head \
+        -m '<Summary of change>'
+
+Where `${branch_name}` is either the name of the extension's module (if the
+script is for an extension) or `haas` (if the script is for HaaS core).
+
+Alembic will generate a migration script in the appropriate directory. Again,
+sanity check the output; Alembic's guesses should not be trusted blindly.
+
+## Writing tests
+
+The file ``tests/unit/migrations.py`` provides some basic infrastructure
+for testing migrations. The comments there describe things in full
+detail, but the basic strategy is, back up a database with a known set
+of objects using ``pg_dump``, and place the result in
+``tests/unit/migrations/``. To generate a dump, execute::
+
+    pg_dump -U $haas_user $database_name --column-inserts > filename.sql
+
+``--column-inserts`` is necessary to ensure that the result can be
+loaded via SQLAlchemy; the default is not valid SQL and takes advantage
+of specific features of the ``psql`` command. You will also need to edit
+the file manually, doing the following:
+
+- If the option lock_timeout is set (``SET lock_timeout = ...``), remove
+  that statement; the version of postgres used by Travis doesn't support
+  it, so it will cause failures. It also isn't required.
+- Delete all statements using the keywords GRANT, REVOKE, or EXTENSION.
+  These will cause permission errors if your database user is not root.
+- Delete all statements of the form ``ALTER TABLE ... SET OWNER TO
+  ...``; these may cause failures if the connection details on the
+  machine which runs the tests are different from yours, since the
+  refer to specific database users/roles.
+- At the top of the file, document the following:
+  - In roughly what stage of HaaS's development this dump was taken,
+    e.g. "Just after re-integrating flask" or "first integration of
+    openvpn"
+  - The commit hash of the revision of HaaS which created the database.
```

Comment:
Quoting henn (2016-04-05 00:46:48)

>    Could you make this "git commit hash". It might seem obvious, but
>    alembic has its own hashes so it'd be good to be explicit.

Done. Fyi, alembic actually *doesn't* have hashes, those are just random
strings. I agree the extra clarity is good though.